feature(mark-as-pushed): MarkAsPushed is now available as a standalone function.

rsdmike opened the pull request:
This PR adds MarkAsPushed as it's own function. This is primarily to support the app service configurable and provides more flexibility for when this action occurs. MarkAsPushed will remain available on the context itself.

BREAKING CHANGE: HTTPPost and MQTTSend no longer automatically call MarkAsPushed upon success. It is upon the developer to ensure the method is called appropriately.

Closes #162

Signed-off-by: Mike Johanson <michael.johanson@intel.com>

lenny-intel commented on README.md:
```diff
@@ -192,30 +193,34 @@ There is one encryption transform included in the SDK that can be added to your
+### CoreData Functions
+These are functions that enable interactions with the CoreData REST API. 
+- `NewCoreData()` - This function returns a `CoreData` instance. This `CoreData` instance is used to access the following function(s).
+  - `MarkAsPushed` - This function provides the MarkAsPushed function from the context as a First-Class Transform that can be called in your pipeline. [See Definition Above](#.MarkAsPushed()). The data passed into this function is passed along since all required information is provided on the context (EventId, CorrelationId,etc.. )
```
`The data passed into this function is passed along since all required information is provided on the context` isn't clear. 
Suggest `No data is passed into this function since all required information is provided on the context`

rsdmike replied:
Data from your pipeline is passed in, you have no control over that, and it is passed along to the next function.

lenny-intel replied:
Ok, i was thinking of the config data. How about  this `The data passed into this function from the pipeline is passed along unmodified`

lenny-intel commented on appcontext/context_test.go:
```diff
@@ -0,0 +1,60 @@
+	ctx := Context{}
+	err := ctx.MarkAsPushed()
+
+	assert.Equal(t, "No EventID or EventChecksum Provided", err.Error())
```
This test will panic is err == nill. Add  assert.NotNil on err and call t.Fatal if returns false

lenny-intel commented on appcontext/context_test.go:
```diff
@@ -0,0 +1,60 @@
+	}
+	err := ctx.MarkAsPushed()
+
+	assert.Contains(t, err.Error(), "http://test/api/v1/event/checksum/checksumValue")
```
Same err == nil issue, Trevor mentioned there are some mocks for core data somewhere that could be used here?

rsdmike replied:
Do you know where/what they are? All i cared about was the correct URL being called, which err will always be not nil when these tests are run in the pipeline as the URL it hits wont be available

lenny-intel replied:
Nope I don't. Ask Trevor.

lenny-intel commented on appcontext/context_test.go:
```diff
@@ -0,0 +1,60 @@
+	}
+	err := ctx.MarkAsPushed()
+
+	assert.Contains(t, err.Error(), "http://test/api/v1/event/id/eventId")
```
Same err == nil issue, also, Trevor mentioned there are some mocks for core data somewhere that could be used here?

lenny-intel commented on pkg/transforms/coredata.go:
```diff
@@ -0,0 +1,24 @@
+
+// NewCoreData Is provided to interact with CoreData
+func NewCoreData() *CoreData {
+	coredata := new(CoreData)
```
QQ: why use of `new` rather than CoreData{} like we have in the others?

rsdmike replied:
Because it was copied from the other tests we have, i.e MqttConfig

lenny-intel replied:
We should be consistent across all the functions and based on this we should be using CoreData{}
https://softwareengineering.stackexchange.com/questions/210399/why-is-there-a-new-in-go

rsdmike committed:
feature(core-data): MarkAsPushed is now available as a standalone function

This PR adds MarkAsPushed as it's own function. This is primarily to support the app service configurable and provides more flexibility for when this action occurs. MarkAsPushed will remain available on the context itself.

BREAKING CHANGE: HTTPPost and MQTTSend no longer automatically call MarkAsPushed upon success. It is upon the developer to ensure the method is called appropriately.

Closes #162

Signed-off-by: Mike Johanson <michael.johanson@intel.com>

rsdmike committed:
Merge branch 'master' into markAsPushed

lenny-goodell reviewed (approved):
LGTM
